### PR TITLE
Support `SDL_WINDOWID`

### PR DESCRIPTION
**Pull request by yunline**
Issue: https://github.com/pygame-community/pygame-ce/issues/1943

Test code:
```python
import tkinter
import threading
import os

root=tkinter.Tk()
root.geometry("640x480")
btn=tkinter.Button(root,text="HelloWorld")
btn.pack()
cv=tkinter.Canvas(root)
cv.pack()

off=0
def pg_loop(hwnd):
    import pygame
    from pygame._sdl2 import video

    os.environ['SDL_WINDOWID']=str(hwnd)

    pygame.display.init()

    sf=pygame.display.set_mode()
    #sf=pygame.display.set_mode(hwnd=hwnd) 

    pos=(20,20)
    while not off:
        sf.fill((0,0,0))
        pygame.draw.rect(sf,(255,0,0),(*pos,10,10))
        pygame.display.update()

        for event in pygame.event.get():
            if event.type==1024:#MOUSEMOTION
                pos=event.pos


if __name__ == "__main__":
    hwnd=cv.winfo_id()
    threading.Thread(target=pg_loop,args=(hwnd,)).start()
    root.mainloop()
    off=1
    
```

### CODE REVIEW

**Comment by yunline** [comment]
Test code for `hwnd` parameter:
```python
import tkinter
import threading

root=tkinter.Tk()
root.geometry("640x480")
btn=tkinter.Button(root,text="HelloWorld")
btn.pack()
cv=tkinter.Canvas(root)
cv.pack()

off=0
def pg_loop(hwnd):
    import pygame
    from pygame._sdl2 import video

    pygame.display.init()

    sf=pygame.display.set_mode(hwnd=hwnd) 

    pos=(20,20)
    while not off:
        sf.fill((0,0,0))
        pygame.draw.rect(sf,(255,0,0),(*pos,10,10))
        pygame.display.update()

        for event in pygame.event.get():
            if event.type==1024:#MOUSEMOTION
                pos=event.pos


if __name__ == "__main__":
    hwnd=cv.winfo_id()
    threading.Thread(target=pg_loop,args=(hwnd,)).start()
    root.mainloop()
    off=1
    
```

**Comment by Starbuck5** [comment]
For the purposes of issue #1943 I was hoping for a more limited PR.

This adds another parameter to set_mode and has complex looking handling of cases where that intersects with OpenGL and different versions of SDL, which I don't understand without doing more research.

**Comment by yunline** [comment]
> For the purposes of issue #1943 I was hoping for a more limited PR.
> 
> This adds another parameter to set_mode and has complex looking handling of cases where that intersects with OpenGL and different versions of SDL, which I don't understand without doing more research.

It's limited now. The rest of contents would be another PR

**Review by MyreMylar — APPROVED** [review]
LGTM 👍

**Comment by ankith26** [comment]
Hmmm, this may actually be a more complex issue.

On my ubuntu machine, the pygame window does embed inside the tk one after this PR, but the application doesn't work as expected (pygame doesn't get any of the events apart from "keymapchanged")

**Comment by MyreMylar** [comment]
> Hmmm, this may actually be a more complex issue.
> 
> On my ubuntu machine, the pygame window does embed inside the tk one after this PR, but the application doesn't work as expected (pygame doesn't get any of the events apart from "keymapchanged")

Dang.

This prompted me to go check events on this PR versus pygame 1.9.6 (when this last worked). 

On Windows
--------------

1.9.6
-----

 -Mouse events - fine
 - keyboard events - fine
 - Seemingly no joystick events (were there any?). 
 
1.9.6 just didn't have as many events as now.


This PR
--------

- Mouse events - fine
- Window events -fine
- **No keyboard events.**
- Also no Joystick events except joy device added (I guess removed as well).

**Comment by oddbookworm** [comment]
I also just checked, and I could trigger every event I tried except keyboard inputs

**Comment by yunline** [comment]
> * No keyboard events.

I solved this by adding two lines to the test code.
I'm trying to figer out why.

```python
import tkinter
import threading
import os

root=tkinter.Tk()
root.geometry("640x480")
btn=tkinter.Button(root,text="HelloWorld")
btn.pack()
cv=tkinter.Canvas(root)
cv.pack()

off=0
def pg_loop(hwnd):
    import pygame
    from pygame._sdl2 import video

    os.environ['SDL_WINDOWID']=str(hwnd)

    pygame.display.init()

    sf=pygame.display.set_mode()
    win=video.Window.from_display_module()
    win.focus()

    pos=(20,20)
    while not off:
        sf.fill((0,0,0))
        pygame.draw.rect(sf,(255,0,0),(*pos,10,10))
        pygame.display.update()

        for event in pygame.event.get():
            if event.type==1024:#MOUSEMOTION
                pos=event.pos

if __name__ == "__main__":
    hwnd=cv.winfo_id()
    threading.Thread(target=pg_loop,args=(hwnd,)).start()
    root.mainloop()
    off=1
```

```
<Event(768-KeyDown {'unicode': ' ', 'key': 32, 'mod': 4096, 'scancode': 44, 'window': <pygame._sdl2.video.Window object at 0x000001A1C7E73270>})>
<Event(769-KeyUp {'unicode': ' ', 'key': 32, 'mod': 4096, 'scancode': 44, 'window': <pygame._sdl2.video.Window object at 0x000001A1C7E73270>})>
<Event(768-KeyDown {'unicode': 'a', 'key': 97, 'mod': 4096, 'scancode': 4, 'window': <pygame._sdl2.video.Window object at 0x000001A1C7E73270>})>
<Event(769-KeyUp {'unicode': 'a', 'key': 97, 'mod': 4096, 'scancode': 4, 'window': <pygame._sdl2.video.Window object at 0x000001A1C7E73270>})>
```

**Comment by oddbookworm** [comment]
It works, until you click out of the tkinter window and back in or press "Tab"

**Comment by yunline** [comment]
> It works, until you click out of the tkinter window and back in or press "Tab"
@oddbookworm 
It seems this is the problem of tkinter.
When running without pygame window, if Tab key were pressed twice, the Canvas would not regain the input focus.

And the docs says:
> Be aware that there can be many strange side effects when running in an embedded display.

So I think this isn't an issue with our implementation.

**Comment by yunline** [comment]
Rerun CI

File: src_c/display.c [hunk 1]
```diff
@@ -847,24 +847,30 @@ pg_set_mode(PyObject *self, PyObject *arg, PyObject *kwds)
...
 
     scale_env = SDL_getenv("PYGAME_FORCE_SCALE");
+    winid_env = SDL_getenv("SDL_WINDOWID");
 
     if (!PyArg_ParseTupleAndKeywords(arg, kwds, "|Oiiii", keywords, &size,
                                      &flags, &depth, &display, &vsync))
         return NULL;
 
+    if (hwnd == 0 && winid_env != NULL) {
+        hwnd = SDL_strtol(winid_env, NULL, 0);
```
Comment (Starbuck5) [inline src_c/display.c:871]: Since you're storing in a long long shouldn't this be SDL_stroll ? (an additional "L" at the end there)

Comment (yunline) [reply src_c/display.c]: Fixed

**Review by Starbuck5 — APPROVED** [review]
So it doesn't "work," and it's an overdue attempt to regain backwards compatibility with a questionable pygame 1 development strategy.

I'm not excited about it, but it looks like we should have this code attempt of compat in here for parity. I don't think there's anything else we can do with the code to make it better. Thank you yunline for working on this!

**Pull request merged**